ext/openssl/openssl.c

```diff
@@ -3566,13 +3610,43 @@ static EVP_PKEY *php_openssl_pkey_from_zval(zval *val, int public_key, char *pas
 				TMP_CLEAN;
 			}
 		}
+		if (Z_STRLEN_P(val) > 7 && memcmp(Z_STRVAL_P(val), "pkcs11:", sizeof("pkcs11:") - 1) == 0) {
+			char *verbose = NULL;
+			engine = ENGINE_by_id("pkcs11");
+			if (engine == NULL) {
+				php_error_docref(NULL, E_WARNING, "Cannot load PKCS11 engine");
+				TMP_CLEAN;
+			}
+			verbose = getenv("OPENSSL_ENGINE_VERBOSE");
+			if (verbose) {
+				if (!ENGINE_ctrl_cmd_string(engine, "VERBOSE", NULL, 0)) {
+					ENGINE_free(engine);
+					TMP_CLEAN;
+				}
+			}
+			if (!ENGINE_init(engine)) {
+				ENGINE_free(engine);
+				ENGINE_finish(engine);
+				engine = NULL;
+				php_error_docref(NULL, E_WARNING, "Cannot init PKCS11 engine");
+				TMP_CLEAN;
+			}
+		}
 		/* it's an X509 file/cert of some kind, and we need to extract the data from that */
 		if (public_key) {
-			cert = php_openssl_x509_from_str(Z_STR_P(val));
+			if (engine) {
+				key = ENGINE_load_public_key(engine, Z_STRVAL_P(val), NULL, NULL);
+				ENGINE_free(engine);
```

Comment:
Couldn't free be done after `ENGINE_init` so it doesn't need to be done here and also in the private key branch?

Reply:
Same, I do not see calls to ENGINE_free() right after ENGINE_init() within `openssl/apps/apps.c/load_key()`. Same, I do not feel confortable.

Some other cases, such as:
  https://lxr.missinglinkelectronics.com/uboot/lib/rsa/rsa-sign.c#L343
always perform some calls to ENGINE_finish(), ENGINE_free() after `ENGINE_by_id()` and `ENGINE_init()` ; which seems to make more sense from the man pages of ENGINE_xyz than the logics from openssl/apps/apps.c ! So I get confused, please, can you comment based on your understanding ?

I'll provide an update.

Reply:
Again just from reading `ENGINE_init` docs and one of the examples (`Issuing control commands to an ENGINE`), the structural can be freed after getting the functional. Also from the docs it seems that finish should free the structural reference anyway so maybe it's not needed at all:

> All functional references are released by calling ENGINE_finish() (which removes the implicit structural reference as well).

Reply:
done: https://github.com/php/php-src/pull/6860/commits/de9cc93182cdccf4f81e52880d42ae669f9a7de6, thanks for your nice review.